lib/protoparser/prometheus/parser.go

```diff
@@ -230,43 +231,101 @@ func unmarshalRow(dst []Row, s string, tagsPool []Tag, noEscapes bool, errLogger
 
 var invalidLines = metrics.NewCounter(`vm_rows_invalid_total{type="prometheus"}`)
 
-func unmarshalTags(dst []Tag, s string, noEscapes bool) (string, []Tag, error) {
+func unmarshalQuotedString(s string, noEscapes bool) (string, string, error) {
+	q := strings.IndexByte(s, '"')
+
+	// Check to see if s can contain 2 characters
+	if q == -1 || len(s[q:]) < 2 {
+		return "", s, fmt.Errorf("missing starting double quote in string: %q", s)
+	}
+	var n int
+	if noEscapes {
+		n = strings.IndexByte(s[q+1:], '"')
+		if n == -1 {
+			return "", s, fmt.Errorf("missing closing double quote in string: %q", s)
+		}
+		// Add 2 to account for both quotes
+		return s[q+1 : q+1+n], s[q+2+n:], nil
+	}
+	n = findClosingQuote(s[q:])
+	if n == -1 {
+		return "", s, fmt.Errorf("missing closing double quote in string: %q", s)
+	}
+	return unescapeValue(s[q+1 : n]), s[n+1:], nil
+}
+
+func (r *Row) unmarshalTags(dst []Tag, s string, noEscapes bool) (string, []Tag, error) {
+	// Edge case for no tags
+	s = skipLeadingWhitespace(s)
+	if s == "" {
+		return s, dst, nil
+	}
+	var err error
+
 	for {
 		s = skipLeadingWhitespace(s)
-		if len(s) > 0 && s[0] == '}' {
-			// End of tags found.
-			return s[1:], dst, nil
-		}
-		n := strings.IndexByte(s, '=')
+		n := strings.IndexByte(s, '"')
 		if n < 0 {
+			// Check that are we done with tags
+			if len(s) > 0 && s[0] == '}' {
+				return s[1:], dst, nil
+			}
 			return s, dst, fmt.Errorf("missing value for tag %q", s)
 		}
-		key := skipTrailingWhitespace(s[:n])
-		if strings.IndexByte(key, '"') >= 0 {
-			return s, dst, fmt.Errorf("tag key %q cannot contain double quotes", key)
+		// Determine if this is a value or quoted label
+		possibleKey := skipTrailingWhitespace(s[:n])
+		possibleKeyLen := len(possibleKey)
+		key := ""
+		if possibleKeyLen == 0 || possibleKey[possibleKeyLen-1] == ',' {
```

Comment:
`possibleKey[possibleKeyLen-1] == ','` is this check needed? Could you please add test case for it?